prov/shm/src/smr_progress.c

```diff
@@ -569,29 +569,59 @@ static int smr_progress_ipc(struct smr_cmd *cmd, enum fi_hmem_iface iface,
 }
 
 static void smr_do_atomic(void *src, void *dst, void *cmp, enum fi_datatype datatype,
-			  enum fi_op op, size_t cnt, uint16_t flags)
+                          enum fi_op op, size_t cnt, uint16_t flags,
+                          enum fi_hmem_iface iface, uint64_t device)
 {
 	char tmp_result[SMR_INJECT_SIZE];
+	char tmp_dst[SMR_INJECT_SIZE];
 
 	if (ofi_atomic_isswap_op(op)) {
-		ofi_atomic_swap_handler(op, datatype, dst, src, cmp,
-					tmp_result, cnt);
+		if (iface == FI_HMEM_SYSTEM) {
+			ofi_atomic_swap_handler(op, datatype, dst, src, cmp, tmp_result, cnt);
+		} else {
+			ofi_copy_from_hmem(iface, device, tmp_dst, dst,
+			                   cnt * ofi_datatype_size(datatype));
+			ofi_atomic_swap_handler(op, datatype, tmp_dst, src, cmp, tmp_result,
+			                        cnt);
+			ofi_copy_to_hmem(iface, device, dst, tmp_dst,
+			                 cnt * ofi_datatype_size(datatype));
```

Comment:
We'll need to decide what it means to be atomic here.  These flows (including below) are only atomic if nothing else is touching the same regions, since we're manually doing read-modify-write.  In these flows, the modify portion doesn't need to use CPU atomics, though it's harmless to keep them.  That begs the question of whether we need CPU atomics in the system memory case.  If we need CPU atomics for system memory, then the non-system memory case would be broken.

Reply:
On the [man page:](https://ofiwg.github.io/libfabric/v1.17.0/man/fi_atomic.3.html)

> The correctness of atomic operations on a target memory region is guaranteed only when performed by a single actor for a given window of time. An actor is defined as a single libfabric domain (identified by the domain name, and not an open instance of that domain), a coherent CPU complex, or other device (e.g. GPU) capable of performing atomic operations on the target memory. The results of atomic operations performed by multiple actors simultaneously are undefined. For example, issuing CPU based atomic operations to a target region concurrently being updated by NIC based atomics may leave the region’s data in an unknown state. The results of a first actor’s atomic operations must be visible to a second actor prior to the second actor issuing its own atomics.


It would appear that we do not need CPU atomics for system memory.

Reply:
I'm assuming the shm provider is using locks to serialize a single process with multiple threads.  However, could 2 different processes access the same GPU memory?  The current atomic implementation to system memory works in this case, as the CPU atomics themselves are used.  In the case of GPU memory, the proposed change may break that.

Example: It looks like shm uses the domain name 'shm' for all domains.  That makes sense.  Supposed an app opens 2 domains, each with access to the same GPU memory region.  If 2 peers perform atomic updates to the GPU memory, but through different domains at the target, can we still be considered atomic?

I think the answer is no, and part of the reason is the use of memcpy.  We could read out the GPU region while another process is writing to the same region.  Even if the memcpy's were atomic, I don't think that's sufficient, since it's possible for both to read out the old value, but only one of the operations would end up being stored in the result.  (Imagine if both wanted to add 1 to the stored value.  The result should be value + 2, but we could end up with value +1.)

I don't know what the answer is here, but it seems like there needs to be some sort of global synchronization between processes when accessing non-system memory.  CPU atomics give us that sort of synchronization.  For GPUs, either we need our own, or we need to schedule the GPU to execute the atomic.

Reply:
What is the reasoning behind defining an actor a libfabric domain identified by name, not instance of domain?

Reply:
It is needed to ensure atomicity from the viewpoint of the remote peer.  Imagine if a domain is a NIC.  All atomic operations that target that NIC are guaranteed to be atomic, regardless of which endpoint is used to access the memory or open software instance.  In this case, the domain name is the name of the NIC.

Reply:
So you are saying that we need an interprocess lock per domain (NIC), per memory region to make our emulated implementation correct?

Reply:
or at the very minimum a 1 atomic per domain lock (if we don't care about performance).

Reply:
Basically, yes, though I'm not sure about per memory region, at least not a libfabric MR.  We somehow need to serialize accesses to the same GPU memory location if we're going to copy the data of to the host, modify it, then move it back to the GPU.

Is there a way to invoke atomic operations on the accelerator itself?  That's what we do with system memory -- call CPU atomics.  This is likely a lot more work, but it's probably what we really want.